### ISSUE TITLE
Can you create a wine-lol with the fsync patch please (futex)?

### ISSUE THREAD

**Issue by Getta91** [M-Reimer/wine-lol#56]
Hi, I love your work there but with the actual wine-lol, I can't use the variable WINEFSYNC = 1. I already have the required kernel but I not the good wine. Can you please make a wine-lol-fsync? I don't know how to proceed (kindanoob :/) so I hope you can help me. Thanks

**Comment by M-Reimer** [comment]
I won't do anything until someone properly fixes this S9 thing. Current state is that I can't play at all.

**Comment by Getta91** [comment]
I understand what you mean. Can you at least help me building it? I am using a linux mint 20 distro but I don't how to proceed.

**Comment by M-Reimer** [comment]
With my scripts you can not build at all on Mint. They are Arch Linux PKGBUILD based. To actually build the DEB files this would require you to set up an Arch virtual machine, first.

**Comment by Getta91** [comment]
oh well i guess i won't get into that stuff then.thanks anyway

**Comment by M-Reimer** [comment]
I'll have a look at this around the end of this month.

**Comment by Getta91** [comment]
Yeah thats a cool thing! Please, prevent us when u give it a shot ;)

**Comment by Grelo4ka** [comment]
> With my scripts you can not build at all on Mint. They are Arch Linux PKGBUILD based. To actually build the DEB files this would require you to set up an Arch or (probably better) Manjaro virtual machine, first.

What do you mean by "(probably better) Manjaro virtual machine"? Do you recommend using Manjaro instead of Arch for running wine-lol?

**Comment by Getta91** [comment]
I think he just meant that manjaro is easier to use when you need the PKGBUILD based thing

**Comment by M-Reimer** [comment]
Exactly. Currently I would suggest Manjaro to noone. If possible use Arch.

**Comment by Grelo4ka** [comment]
> I think he just meant that manjaro is easier to use when you need the PKGBUILD based thing.
> By the way, I am using manjaro right now. If you could tell me how to build it, That would be very kind.

You can easily build wine-lol on Manjaro from AUR but you need an AUR helper for this. For pamac, the command would be `pamac install wine-lol`, and that's it. 
Make sure to enable multilib repos in `/etc/pacman.conf` if you haven't done it yet (uncomment 2 lines with multilib, then do `paman -Syu`).
To enable AUR in pamac, uncomment "EnableAUR" in `/etc/pamac.conf` (or use this command: `sudo sed --in-place "s/#EnableAUR/EnableAUR/" "/etc/pamac.conf"`).
Please note that it will take time to build (30 minutes on my PC).

**Comment by M-Reimer** [comment]
He needs a build for a Debian derivative. And that's not that easy.

And about Manjaro (even if Off-Topic): https://forum.manjaro.org/t/change-of-treasurer-for-manjaro-community-funds/154888

I have more time recently as my exams are written. Still have to go through some other work but I'll come back to this one and do a test build with the fsync patch.

**Comment by M-Reimer** [comment]
I'll give up on this topic for now.

Seems like since wine 5.10 even esync is disabled and fsync doesn't apply at all. It would be possible to hack something into wine 5.6 but then we are basically locked to this version as everything from 5.7 to 5.10 does not launch LoL at all.

For me it even seems like Valve has given up on fsync. No work in that direction...

**Comment by kassindornelles** [comment]
Valve didn't gave up of FSYNC, they are just working to get the kernels patches in a state that is good for upstream, it takes time.
anyway, you could easily add ESYNC that does not need any kernel mode to work, is fully user-space
and it works on wine 5.7

**Comment by yuiiio** [comment]
I success fsync with wine-5.6
https://github.com/yuiiio/wine-lol-src
this repo include chosen patches from staging to play lol
and fsync-esync(but maybe esync is unusable because eventfd limit) commit from valve

plus
HACK: Don't build winemenubuilder (from valve)
change OpenCL to CL
Improve heap allocation performance (from satging)
Never use CLOCK_MONOTONIC_RAW (from valve)

can test like
https://github.com/yuiiio/wine-lol

**Comment by M-Reimer** [comment]
For Wine 5.6 it should be possible to just apply one patch (just have to find the right one. There are a few of them for different versions of Wine).

But the problem with this is that, once it gets possible again to upgrade to a more recent version, fsync will be gone again. That's one of the big reasons why I don't want to enable it now.

I really hope that, once the required change is integrated into the kernel, this new feature will be implemented directly in Wine(-staging) and no more manual patching is needed.

**Comment by yuiiio** [comment]
you are right.
extra patches will be harder to maintain.

**Comment by yuiiio** [comment]
If you're interested. I added Fsync patch from Tkg and staging+fix in 5.9 top of m-reimer/master 
https://github.com/yuiiio/wine-lol/commits/master

**Comment by M-Reimer** [comment]
Actually I don't see a big need to move from 5.6 (old version) to 5.9 (another old version).
As we can't go to anything more "up to date", I would prefer to "never change a running system" for now.

**Comment by yuiiio** [comment]
I pulled out the patches needed to make the lol work in staging(5.9).https://github.com/yuiiio/wine-lol-src/commits/slim

I've applied those patches on top of 5.18 (the many ntdll reverts from TKG).https://github.com/yuiiio/wine-lol-src/commits/wine-new-work
It seems to work.

**Comment by yuiiio** [comment]
I noticed that some patches were missed during the tkg rebase.

**Comment by yuiiio** [comment]
update to 5.18 + tkg's fsync + tkg's missing patch
https://github.com/yuiiio/wine-lol/commit/ea9c74c2d807d17e428aff8438e63911ef9f2341

**Comment by M-Reimer** [comment]
Marking this one WONTFIX now.

Reason: "Futex2" is in the works and once this is finished it will be hopefully available directly in one of the future stable kernels. Once this happens it is very likely that WINE itself will support this "official kernel feature" at some time.

The current state with "futex" is, that literally dozens of commits in a current WINE version have to be reverted to get back to a state where the patch still applies. I don't want to deal with that.
